File: 2nd year/MPP/project-mpp/java-project/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepository/DBRefereeRepository.java
package triathlon.persistence.dbRepository;

import triathlon.model.Game;
import triathlon.model.Referee;
import triathlon.persistence.RefereeRepository;
import triathlon.persistence.RepositoryException;
import triathlon.persistence.validators.ValidationException;
import triathlon.persistence.validators.Validator;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class DBRefereeRepository implements RefereeRepository {
    private final Validator<Referee> validator;
    private final Connection connection;

    public DBRefereeRepository(Properties props, Validator<Referee> validator) {
        this.connection = new JDBCManager(props).getConnection();
        this.validator = validator;
    }

    @Override
    public Referee findOne(Long id) {
        final String sql_query = "select r.id as id, g.id as gid, r.name as name, g.name as gname, username, password " +
                "from Referee r, Game g where id = ? and r.gameId=g.id";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                Referee referee = new Referee(
                        resultSet.getLong("id"),
                        new Game(resultSet.getLong("gid"), resultSet.getString("gname")),
                        resultSet.getString("name"),
                        resultSet.getString("username"),
                        resultSet.getString("password")
                );
                return referee;
            }
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public Iterable<Referee> findAll() {
        final String sql_query = "select r.id as id, g.id as gid, r.name as name, g.name as gname, username, password" +
                " from Referee r, Game g where r.gameId=g.id";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            List<Referee> referees = new ArrayList<>();
            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                referees.add(new Referee(
                        resultSet.getLong("id"),
                        new Game(resultSet.getLong("gid"), resultSet.getString("gname")),
                        resultSet.getString("name"),
                        resultSet.getString("username"),
                        resultSet.getString("password")
                ));
            }
            return referees;
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }

        return null;
    }

    @Override
    public void save(Referee referee) {
        final String sql_query =
                "insert into Referee (gameId, name, username, password) values (?,?,?,?)";
        try {
            validator.validate(referee);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, referee.getGame().getId());
            statement.setString(2, referee.getName());
            statement.setString(3, referee.getUsername());
            statement.setString(4, referee.getPassword());

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void update(Referee referee) {
        final String sql_query = "update Referee set gameId = ?, name = ?, username = ?, password = ? where id = ?";
        try {
            validator.validate(referee);
        } catch (ValidationException ex) {
            throw new RepositoryException(ex.getMessage());
        }

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, referee.getGame().getId());
            statement.setString(2, referee.getName());
            statement.setString(3, referee.getUsername());
            statement.setString(4, referee.getPassword());
            statement.setLong(5, referee.getId());

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }

    @Override
    public void delete(Long id) {
        final String sql_query = "delete from Referee where id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql_query)) {
            statement.setLong(1, id);

            int updates = statement.executeUpdate();
        } catch (SQLException e) {
            System.err.println("Error DB: " + e);
        }
    }
}

File: 2nd year/MPP/project-mpp/csharp-project/persistence/dbRepository/DbRefereeRepository.cs
using System;
using System.Collections.Generic;
using log4net;
using Microsoft.Data.Sqlite;
using model;
using persistence.validators;

namespace persistence.dbRepository
{
    public class DbRefereeRepository : IRefereeRepository
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(DbRefereeRepository));
        private readonly SqliteConnection _connection = JdbcManager.GetConnection();
        private readonly IValidator<Referee> _validator;

        public DbRefereeRepository(IValidator<Referee> validator)
        {
            this._validator = validator;
        }

        public Referee FindOne(long id)
        {
            Log.Info($"Entered with id = {id}");
            try
            {
                const string commandText = "select * from Referee where id = @id";
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                using var reader = command.ExecuteReader();
                if (reader.Read())
                {
                    return new Referee
                    {
                        Id = reader.GetInt64(0),
                        Game = new Game
                        {
                            Id = reader.GetInt64(1)
                        },
                        Name = reader.GetString(2),
                        Username = reader.GetString(3),
                        Password = reader.GetString(4)
                    };
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            return null;
        }

        public IEnumerable<Referee> FindAll()
        {
            Log.Info("");
            const string commandText = "select * from Referee";
            IList<Referee> referees = new List<Referee>();

            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                using var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    referees.Add(new Referee
                    {
                        Id = reader.GetInt64(0),
                        Game = new Game
                        {
                            Id = reader.GetInt64(1)
                        },
                        Name = reader.GetString(2),
                        Username = reader.GetString(3),
                        Password = reader.GetString(4)
                    });
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }

            Log.Info(referees);
            return referees;
        }

        public void Save(Referee referee)
        {
            Log.Info($"Entered with entity = {referee}");
            const string commandText =
                "insert into Referee (id, gameId, name, username, password) values (@id, @gameId, @name, @username, @password)";
            _validator.Validate(referee);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", referee.Id);
                command.Parameters.AddWithValue("gameId", referee.Game.Id);
                command.Parameters.AddWithValue("name", referee.Name);
                command.Parameters.AddWithValue("username", referee.Username);
                command.Parameters.AddWithValue("password", referee.Password);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Update(Referee referee)
        {
            Log.Info($"Entered with entity = {referee}");
            const string commandText =
                "update Referee set gameId = @gameId, name = @name, username = @username, password = @password where id = @id";
            _validator.Validate(referee);
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", referee.Id);
                command.Parameters.AddWithValue("gameId", referee.Game.Id);
                command.Parameters.AddWithValue("name", referee.Name);
                command.Parameters.AddWithValue("username", referee.Username);
                command.Parameters.AddWithValue("password", referee.Password);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }

        public void Delete(long id)
        {
            Log.Info($"Entered with id = {id}");
            const string commandText = "delete from Referee where id = @id";
            try
            {
                using var command = new SqliteCommand(commandText, _connection);
                Log.Debug(commandText);

                command.Parameters.AddWithValue("id", id);

                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                Log.Error(ex.Message);
                Console.Write(ex.Message);
            }
        }
    }
}